File: web/src/lib/image/colorUtils.ts
export type RGBA = [number, number, number, number];

export function relativeLuminance([r, g, b]: RGBA): number {
  return 0.2126 * r + 0.7152 * g + 0.0722 * b;
}

export function toCssColor([r, g, b]: RGBA): string {
  return `rgb(${Math.round(r)}, ${Math.round(g)}, ${Math.round(b)})`;
}

export function contrastColorFor([r, g, b]: RGBA): RGBA {
  return relativeLuminance([r, g, b, 255]) > 140 ? [0, 0, 0, 255] : [255, 255, 255, 255];
}

export function medianColor(samples: RGBA[]): RGBA {
  if (samples.length === 0) return [128, 128, 128, 255];
  const channel = (i: number) => {
    const values = samples.map((s) => s[i]).sort((a, b) => a - b);
    return values[Math.floor(values.length / 2)];
  };
  return [channel(0), channel(1), channel(2), channel(3)];
}

export function colorDistance(a: RGBA, b: RGBA): number {
  return Math.sqrt((a[0] - b[0]) ** 2 + (a[1] - b[1]) ** 2 + (a[2] - b[2]) ** 2);
}

File: android/app/src/main/java/com/noctis/app/image/ColorUtils.kt
package com.noctis.app.image

import android.graphics.Color

fun relativeLuminance(color: Int): Double =
    0.2126 * Color.red(color) + 0.7152 * Color.green(color) + 0.0722 * Color.blue(color)

fun contrastColorFor(bgColor: Int): Int = if (relativeLuminance(bgColor) > 140) Color.BLACK else Color.WHITE

fun medianColor(samples: List<Int>): Int {
    if (samples.isEmpty()) return Color.GRAY
    fun channelMedian(selector: (Int) -> Int): Int {
        val sorted = samples.map(selector).sorted()
        return sorted[sorted.size / 2]
    }
    return Color.rgb(
        channelMedian { Color.red(it) },
        channelMedian { Color.green(it) },
        channelMedian { Color.blue(it) },
    )
}

fun colorDistance(a: Int, b: Int): Double {
    val dr = (Color.red(a) - Color.red(b)).toDouble()
    val dg = (Color.green(a) - Color.green(b)).toDouble()
    val db = (Color.blue(a) - Color.blue(b)).toDouble()
    return Math.sqrt(dr * dr + dg * dg + db * db)
}
